src/main/java/org/opentripplanner/routing/algorithm/raptoradapter/transit/constrainedtransfer/ConstrainedBoardingSearch.java

```diff
@@ -105,9 +105,15 @@ public boolean transferExist(int targetStopPos) {
             int stopIndex
     ) {
         final Trip trip = tripSchedule.getOriginalTripTimes().getTrip();
-        return currentTransfers.stream()
-                .filter(t -> t.matchesSourcePoint(stopIndex, trip))
-                .collect(Collectors.toList());
+        // for performance reasons we use a for loop here as streams are much slower.
+        // i experimented with LinkedList and ArrayList and LinkedList was faster.
+        var result = new LinkedList<TransferForPattern>();
```

Comment:
As @hannesj pointed out using LinkedList might be dangerous, while appending to it is fast - random access is not. The list is returned, so you have no control how it will be used - I suggest sticking with ArrayList. If we want to optimize this, then reusing the ArrayList is probably better(not looked if it is even possible).

Reply:
What if I returned a `Collection` instead in order to communicate that you should only ever iterate over it rather than access stuff by index?